### ISSUE TITLE
Diagnostic for use of plain protocol name in type position should suggest `some` or `any`

### ISSUE THREAD

**Issue by hborla** [swiftlang/swift#68284]
The following code is invalid:

```swift
protocol P {
  associatedtype A
}

func generic(value: P) {}
```

The compiler currently produces the error message `Use of protocol 'P' as a type must be written 'any P'` with a fix-it to insert the `any` keyword.

However, a different (and often better!) fix here is to use the `some` keyword. The error message should be re-worded to prompt the programmer to consider `any` or `some`, and there should be two notes, each with a fix-it to insert `any` or `some`, respectively.

**Comment by CrazyFanFan** [comment]
I would like to solve this problem, Do you mind if I start?

**Comment by Dhandeep10** [comment]
@hborla 
sir, what is meant here, can you please elaborate, so that I can contribute accordingly.

**Comment by taevonlewis** [comment]
I’m currently working on fixing this bug in diagnostics as part of the swift mentorship program

**Comment by Dhandeep10** [comment]
I too want to contribute and get started, I am new to open source, so want to learn and explore this.

**Comment by saehejkang** [comment]
@taevonlewis are you still working on this issue?

**Comment by li3zhen1** [comment]
Hi @taevonlewis, are you still working on this?

**Comment by saehejkang** [comment]
@AnthonyLatsis can this be reassigned due to its lack of activity?

**Comment by AnthonyLatsis** [comment]
`-verify` mode diagnoses both grammatical issues and mismatches — with details on actual vs. expected diagnostics.

**Comment by AnthonyLatsis** [comment]
@taevonlewis Are you fine with passing this over to Saehej?

**Comment by taevonlewis** [comment]
Yes, they can take over this issue

**Comment by saehejkang** [comment]
> with details on actual vs. expected diagnostics (fix-its).

With the `-verify` flag added to the `swift-frontend` build arguments, I get this output (no details on actual), just an error that the note is not produced.

```
/Documents/xcode/SwiftCompiler/SwiftCompiler/main.swift:71:28: error: expected note not produced
                        // expected-note@-1 {{Replace 'P' with 'any P'}} {{21-21=any P}}
~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by AnthonyLatsis** [comment]
Right, there is no actual note emitted on the specified line (and at all), so there is nothing to compare your expectation with.

**Comment by saehejkang** [comment]
> there is no actual note emitted on the specified line

<img width="362" alt="Screenshot 2024-05-01 at 10 36 58 AM" src="https://github.com/apple/swift/assets/20051028/3aaaa229-ee7f-4ad0-ab1a-d151b30c52e0">

Is the `fix-it` not an expected note? Maybe I wrote one of these wrong

```
// expected-error {{use of protocol 'P' as a type must be written 'any P'}}
// expected-note@-1 {{Replace 'P' with 'any P'}} {{21-21= any P}}
```

**Comment by saehejkang** [comment]
We want to make updates to this diagnostic I found below. I was thinking to add a new variable `Type` and pass in the `'some P'` to the error, but I had some questions about where that comes from. 

```
ERROR(existential_requires_any,none, "use of %select{protocol |}2%0 as a type must be written %1", 
(Type, Type, bool))
```

On this line in [TypeCheckType](https://github.com/apple/swift/blob/5cbf815f22aebd3f2520f6210f589826b65dacb8/lib/Sema/TypeCheckType.cpp#L6124) the diagnostic is used and passes in the `Type` returned by `getDeclaredExistentialType()`. Correct me if I am wrong, but I believe that is the value `'any P'`. The `ExistentialType` is spelled with the keyword `any` as it says so here in [Types](https://github.com/apple/swift/blob/5cbf815f22aebd3f2520f6210f589826b65dacb8/include/swift/AST/Types.h#L6227). 

My thought process was to look for something similar that uses the keyword `some` and I landed on looking into `OpaqueTypes`. Is there something similar to `getDeclaredExistentialType()` that pretty much fetches the value `'some P'`?  I delved deeper into `OpaqueTypes` and feel I need use something like [this](https://github.com/apple/swift/blob/5cbf815f22aebd3f2520f6210f589826b65dacb8/include/swift/AST/Types.h#L6579) .

**Comment by AnthonyLatsis** [comment]
@hborla Do you reckon we should also suggest both options in positions where `some X` is an opaque result type? Also, we should probably refrain from suggesting `some` if the parent declaration is ABI-public and library evolution is enabled. That could cause an ABI break.

**Comment by saehejkang** [comment]
@AnthonyLatsis I won't be able to work on this issue anymore. I noticed you merged changes that streamlined the logic in the syntax checker. Unfortunately, I do not have the time to redo the logic based on your changes.

**Comment by AnthonyLatsis** [comment]
No worries, thanks for letting me know.

**Comment by yyoungbash** [comment]
Can you please assign me? @AnthonyLatsis
